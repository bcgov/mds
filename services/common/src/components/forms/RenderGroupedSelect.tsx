import React from "react";
import PropTypes from "prop-types";
import { Select, Form } from "antd";

/**
 * @constant RenderGroupedSelect - Ant Design `Select` component for redux-form - used for data sets that require grouping.
 * There is a bug when the data sets are large enough to cause the dropdown to scroll, and the field is in a modal.
 * In the case where the modal cannot scroll, it is better to pass in the prop doNotPinDropdown.  It allows the
 * dropdown to render properly
 */

const propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  input: PropTypes.objectOf(PropTypes.any).isRequired,
  placeholder: PropTypes.string,
  label: PropTypes.string,
  meta: PropTypes.any, //CustomPropTypes.formMeta,
  data: PropTypes.any, //CustomPropTypes.groupOptions,
  disabled: PropTypes.bool,
  onSelect: PropTypes.func,
  usedOptions: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)),
};

const defaultProps = {
  placeholder: "",
  label: "",
  data: [],
  disabled: false,
  meta: {},
  onSelect: () => {},
  usedOptions: [],
};

const RenderGroupedSelect = (props) => (
  <Form.Item
    label={props.label}
    validateStatus={
      props.meta.touched ? (props.meta.error && "error") || (props.meta.warning && "warning") : ""
    }
    help={
      props.meta.touched &&
      ((props.meta.error && <span>{props.meta.error}</span>) ||
        (props.meta.warning && <span>{props.meta.warning}</span>))
    }
  >
    <Select
      virtual={false}
      disabled={props.disabled}
      dropdownMatchSelectWidth
      showSearch
      placeholder={props.placeholder}
      optionFilterProp="children"
      id={props.id}
      defaultValue={props.input.value}
      value={props.input.value ? props.input.value : undefined}
      onChange={props.input.onChange}
      onSelect={props.onSelect}
    >
      {props.data.map((group) => (
        <Select.OptGroup label={group.groupName} key={group.groupName}>
          {group.opt.map((opt) => (
            <Select.Option
              disabled={props.usedOptions && props.usedOptions.includes(opt.value)}
              key={opt.value}
              value={opt.value}
            >
              {opt.label}
            </Select.Option>
          ))}
        </Select.OptGroup>
      ))}
    </Select>
  </Form.Item>
);

RenderGroupedSelect.propTypes = propTypes;
RenderGroupedSelect.defaultProps = defaultProps;

export default RenderGroupedSelect;
