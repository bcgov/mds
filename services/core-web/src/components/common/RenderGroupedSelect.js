import React from "react";
import PropTypes from "prop-types";
import { Form, Select } from "antd";
import CustomPropTypes from "@/customPropTypes";

/**
 * @constant RenderGroupedSelect - Ant Design `Select` component for redux-form - used for data sets that require grouping.
 * There is a bug when the data sets are large enough to cause the dropdown to scroll, and the field is in a modal.
 * In the case where the modal cannot scroll, it is better to pass in the prop doNotPinDropdown.  It allows the
 * dropdown to render properly
 */

const propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  input: PropTypes.objectOf(PropTypes.any).isRequired,
  placeholder: PropTypes.string,
  label: PropTypes.string,
  meta: CustomPropTypes.formMeta,
  data: CustomPropTypes.groupOptions,
  disabled: PropTypes.bool,
  onSelect: PropTypes.func,
  doNotPinDropdown: PropTypes.bool,
};

const defaultProps = {
  placeholder: "",
  label: "",
  data: [],
  disabled: false,
  meta: {},
  onSelect: () => {},
  doNotPinDropdown: false,
};

const RenderGroupedSelect = (props) => (
  <Form.Item
    label={props.label}
    validateStatus={
      props.meta.touched ? (props.meta.error && "error") || (props.meta.warning && "warning") : ""
    }
    help={
      props.meta.touched &&
      ((props.meta.error && <span>{props.meta.error}</span>) ||
        (props.meta.warning && <span>{props.meta.warning}</span>))
    }
  >
    <Select
      disabled={props.disabled}
      getPopupContainer={props.doNotPinDropdown ? null : () => document.getElementById(props.id)}
      showSearch
      placeholder={props.placeholder}
      optionFilterProp="children"
      filterOption={(input, option) =>
        typeof option.props.children === "string" &&
        option.props.children.toLowerCase().includes(input.toLowerCase())
      }
      id={props.id}
      {...props.input}
      onSelect={props.onSelect}
    >
      {props.data.map((group) => (
        <Select.OptGroup label={group.groupName} key={group.groupName}>
          {group.opt.map((opt) => (
            <Select.Option
              disabled={props.usedOptions && props.usedOptions.includes(opt.value)}
              key={opt.value}
              value={opt.value}
            >
              {opt.label}
            </Select.Option>
          ))}
        </Select.OptGroup>
      ))}
    </Select>
  </Form.Item>
);

RenderGroupedSelect.propTypes = propTypes;
RenderGroupedSelect.defaultProps = defaultProps;

export default RenderGroupedSelect;
