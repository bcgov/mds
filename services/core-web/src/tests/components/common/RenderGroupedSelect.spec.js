import React from "react";
import { shallow } from "enzyme";
import RenderGroupedSelect from "@/components/common/RenderGroupedSelect";

let props = {};

const setupProps = () => {
  props = {
    input: { value: "", onChange: jest.fn() },
    label: "",
    meta: {
      touched: false,
      error: false,
      warning: false,
    },
    id: "parties",
    disabled: false,
    data: [],
    option: {},
    selectedOption: { key: "1" },
    handleChange: jest.fn(),
    usedOptions: {},
  };
};

beforeEach(() => {
  setupProps();
});

describe("RenderGroupedSelect", () => {
  it("renders properly", () => {
    const wrapper = shallow(<RenderGroupedSelect {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
