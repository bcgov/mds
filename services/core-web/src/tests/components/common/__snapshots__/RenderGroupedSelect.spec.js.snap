// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderGroupedSelect renders properly 1`] = `
<FormItem
  hasFeedback={false}
  help={false}
  label=""
  validateStatus=""
>
  <Select
    bordered={true}
    choiceTransitionName=""
    defaultValue=""
    disabled={false}
    dropdownMatchSelectWidth={true}
    id="parties"
    onChange={[MockFunction]}
    onSelect={[Function]}
    optionFilterProp="children"
    placeholder=""
    showSearch={true}
    transitionName="slide-up"
    virtual={false}
  />
</FormItem>
`;
